### PR TITLE
[SPARK-34346][CORE][SQL] io.file.buffer.size set by spark.buffer.size will override by loading hive-site.xml accidentally may cause perf regression

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

In many real-world cases, when interacting with hive catalog through Spark SQL, users may just share the `hive-site.xml` for their hive jobs and make a copy to `SPARK_HOME`/conf w/o modification. In Spark, when we generate Hadoop configurations, we will use `spark.buffer.size(65536)` to reset `io.file.buffer.size(4096)`. But when we load the hive-site.xml, we may ignore this behavior and reset `io.file.buffer.size` again according to `hive-site.xml`.

1. The configuration priority for setting Hadoop and Hive config here is not right, while literally, the order should be `spark > spark.hive > spark.hadoop > hive > hadoop`

2. This breaks `spark.buffer.size` congfig's behavior for tuning the IO performance w/ HDFS if there is an existing `io.file.buffer.size` in hive-site.xml

### Why are the changes needed?

bugfix for configuration behavior and fix performance regression by that behavior change
 
### Does this PR introduce _any_ user-facing change?

this pr restores silent user face change

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @dongjoon-hyun @HyukjinKwon thanks

**Comment by srowen** [comment]
spark > spark.hive > spark.hadoop > hive > hadoop makes sense to me.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 1]
```diff
@@ -1151,6 +1150,37 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
     assert(sc.listJars().exists(_.contains("org.apache.hive_hive-storage-api-2.7.0.jar")))
     assert(sc.listJars().exists(_.contains("commons-lang_commons-lang-2.6.jar")))
   }
+
+  test("SPARK-34346: hadoop configuration priority for spark/hive/hadoop configs") {
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala:1154]: For good measure, do we want to also test that the default io.file.buffer.size you get from a plain Spark configuration is in fact 65536?

Comment (yaooqinn) [reply core/src/test/scala/org/apache/spark/SparkContextSuite.scala:1154]: I guess that will cause test flakiness if it gets a nondefault value somewhere due to the non-deterministic test order in CIs. I did a pre-check for loading hive-site.xml explicitly to ensure it gets loaded but not pollute the final result.

Comment (yaooqinn) [reply core/src/test/scala/org/apache/spark/SparkContextSuite.scala:1154]: oh, I can just set it explicitly too

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfDriverFeatureStepSuite.scala [hunk 2]
```diff
@@ -29,6 +29,7 @@ import org.apache.spark.deploy.k8s._
 import org.apache.spark.deploy.k8s.Constants._
 import org.apache.spark.util.{SparkConfWithEnv, Utils}
 
+
```
Comment (cloud-fan) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfDriverFeatureStepSuite.scala:32]: unnecessary change

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 3]
```diff
@@ -220,31 +219,16 @@ object SharedState extends Logging {
   }
 
   /**
-   * Load hive-site.xml into hadoopConf and determine the warehouse path we want to use, based on
-   * the config from both hive and Spark SQL. Finally set the warehouse config value to sparkConf.
+   * Determine the warehouse path by spark conf, hadoop configuration and the initial options from
+   * the very first created SparkSession instance.
    */
-  def loadHiveConfFile(
+  def determineWarehouse(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:225]: why the return type is still a `Map`?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:225]: We should document what this method returns, as it's not clear from the name.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 4]
```diff
@@ -1151,6 +1150,39 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
...
+
+    resetSparkContext()
+
+    sparkConf
+      .set("spark.hadoop.hadoop.tmp.dir", "/tmp/hive_two")
+      .set(s"spark.hadoop.$bufferKey", "20181117")
+    sc = new SparkContext(sparkConf)
+    assert(sc.hadoopConfiguration.get(testKey) === "/tmp/hive_two",
+      "spark.hadoop configs have higher priority than hive/hadoop ones")
+    assert(sc.hadoopConfiguration.get(bufferKey).toInt === 65536,
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala:1183]: shouldn't this be `20181117`?

Comment (yaooqinn) [reply core/src/test/scala/org/apache/spark/SparkContextSuite.scala:1183]: here is the same sparkConf w/ two more configs，willnot change to respect spark.hadoop.xxx

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:225]: make sense to me. the returned map is unrelated to this change，I will doc better

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala (hunk 4 above)
Comment (cloud-fan) [reply core/src/test/scala/org/apache/spark/SparkContextSuite.scala:1183]: OK, so for the buffer size, we ignore everything but only respect the `BUFFER_SIZE` config.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SharedStateSuite.scala [hunk 5]
```diff
@@ -52,15 +52,4 @@ class SharedStateSuite extends SharedSparkSession {
     assert(conf.isInstanceOf[Configuration])
     assert(conf.asInstanceOf[Configuration].get("fs.defaultFS") == "file:///")
   }
-
-  test("SPARK-33740: hadoop configs in hive-site.xml can overrides pre-existing hadoop ones") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SharedStateSuite.scala:56]: Is this removed because this is merged to the new test coverage?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 6]
```diff
@@ -220,31 +219,16 @@ object SharedState extends Logging {
...
-
     val hiveWarehouseKey = "hive.metastore.warehouse.dir"
-    val configFile = Utils.getContextOrSparkClassLoader.getResourceAsStream("hive-site.xml")
-    if (configFile != null) {
-      logInfo(s"loading hive config file: $configFile")
-      val hadoopConfTemp = new Configuration()
-      hadoopConfTemp.clear()
-      hadoopConfTemp.addResource(configFile)
-      for (entry <- hadoopConfTemp.asScala if !containsInSparkConf(entry.getKey)) {
-        hadoopConf.set(entry.getKey, entry.getValue)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:245]: This removal seems to have some side-effect. Is it okay?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:245]: This behavior exists before SPARK-33740. So, I'm curious about the side-effect.

File: core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala [hunk 7]
```diff
@@ -450,13 +450,22 @@ private[spark] object SparkHadoopUtil {
           hadoopConf.set("fs.s3a.session.token", sessionToken)
         }
       }
+      loadHiveConfFile(conf, hadoopConf)
       appendSparkHadoopConfigs(conf, hadoopConf)
       appendSparkHiveConfigs(conf, hadoopConf)
       val bufferSize = conf.get(BUFFER_SIZE).toString
       hadoopConf.set("io.file.buffer.size", bufferSize)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:457]: Hi, @MaxGekk . According to your email, is this the only property affected?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala (hunk 6 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:245]: According to the current usage restrictions of Hive in Spark, for documented behaviors, there is no side-effect that makes practical sense. But in some undocumented areas, there do have some kind of side effects, e.g. dynamically load the `hive-site.xml` which is unreachable at the start of a Spark app, but added later through some APIs, then those configurations will be added anymore.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SharedStateSuite.scala (hunk 5 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/SharedStateSuite.scala:56]: yes, the newly added case will take over this one

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:245]: Loading `hive-site.xml` dynamically at runtime is really hacky, I don't think anyone would rely on that...

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
`spark > spark.hive > spark.hadoop > hive > hadoop` makes sense and the new change is much easier to prove that it follows this order. Previously, we load the `hive-site.xml` at the end, and try to simulate that it has a lower priority, which is very hacky.

**Comment by cloud-fan** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
Hi @dongjoon-hyun , do you have more concerns about this fix?

File: core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala (hunk 7 above)
Comment (MaxGekk) [reply core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:457]: @dongjoon-hyun Yes.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @yaooqinn , @srowen , @cloud-fan , @MaxGekk .

cc @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.1.

**Comment by HyukjinKwon** [comment]
@yaooqinn it has a conflict in the test. Do you mind opening a backporting PR? I would also like to make sure the tests pass before merging it in since we're very close to the release.

**Comment by yaooqinn** [comment]
OK, it's my pleasure

**Comment by dongjoon-hyun** [comment]
This seems to make the following UT flaky in both GitHub Action and Jenkins.
```
YarnClusterSuite.yarn-cluster should respect conf overrides in SparkHadoopUtil (SPARK-16414, SPARK-23630)
org.scalatest.exceptions.TestFailedException: FAILED did not equal FINISHED (stdout/stderr was not captured)
```

**Comment by dongjoon-hyun** [comment]
Even in this PR, the last commit has the same failure in GitHub Action.

**Comment by HyukjinKwon** [comment]
@yaooqinn would you mind taking a look please?

**Comment by dongjoon-hyun** [comment]
It turns out `core/src/test/resources/core-site.xml` are loading in YARN test. I'll make a fix for this.

**Comment by dongjoon-hyun** [comment]
I made a follow-up here. It switched the way of testing by replacing `core-site.xml` with an ad-hoc `set` command.
- https://github.com/apache/spark/pull/31515
